File: nestjs-api/src/services/analytics.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Order } from '../models/order.entity';
import { Product } from '../models/product.entity';

@Injectable()
export class AnalyticsService {
  constructor(
    @InjectRepository(Order)
    private orderRepository: Repository<Order>,
    @InjectRepository(Product)
    private productRepository: Repository<Product>,
  ) {}

  async getOrdersWithUsers(limit: number = 10, offset: number = 0) {
    return this.orderRepository.createQueryBuilder('o')
      .leftJoin('o.user', 'u')
      .leftJoin('o.orderItems', 'oi')
      .select('o.id', 'order_id')
      .addSelect('o.user_id', 'user_id')
      .addSelect('u.name', 'user_name')
      .addSelect('u.email', 'user_email')
      .addSelect('u.city', 'user_city')
      .addSelect('o.total_amount', 'total_amount')
      .addSelect('o.status', 'status')
      .addSelect('o.order_date', 'order_date')
      .addSelect('COUNT(oi.id)', 'item_count')
      .groupBy('o.id')
      .addGroupBy('o.user_id')
      .addGroupBy('u.name')
      .addGroupBy('u.email')
      .addGroupBy('u.city')
      .addGroupBy('o.total_amount')
      .addGroupBy('o.status')
      .addGroupBy('o.order_date')
      .orderBy('o.id', 'ASC')
      .limit(limit)
      .offset(offset)
      .getRawMany();
  }

  async getUserOrderSummary(limit: number = 10, offset: number = 0) {
    return this.orderRepository.createQueryBuilder('o')
      .leftJoin('o.user', 'u')
      .select('u.id', 'user_id')
      .addSelect('u.name', 'user_name')
      .addSelect('u.email', 'user_email')
      .addSelect('COUNT(o.id)', 'total_orders')
      .addSelect('COALESCE(SUM(o.total_amount), 0)', 'total_amount')
      .addSelect('COALESCE(AVG(o.total_amount), 0)', 'average_order')
      .addSelect('COALESCE(MAX(o.order_date), \'1970-01-01\')', 'last_order')
      .groupBy('u.id')
      .addGroupBy('u.name')
      .addGroupBy('u.email')
      .orderBy('total_amount', 'DESC')
      .limit(limit)
      .offset(offset)
      .getRawMany();
  }

  async getComplexAnalytics() {
    return this.productRepository.createQueryBuilder('p')
      .leftJoin('p.orderItems', 'oi')
      .leftJoin('oi.order', 'o')
      .leftJoin('o.user', 'u')
      .select('p.category', 'category')
      .addSelect('COUNT(DISTINCT o.id)', 'total_orders')
      .addSelect('SUM(oi.quantity)', 'total_quantity')
      .addSelect('SUM(oi.price * oi.quantity)', 'total_revenue')
      .addSelect('AVG(oi.price)', 'avg_price')
      .addSelect('COUNT(DISTINCT u.id)', 'unique_customers')
      .addSelect('AVG(u.age)', 'avg_customer_age')
      .where('o.status = :status', { status: 'completed' })
      .groupBy('p.category')
      .orderBy('total_revenue', 'DESC')
      .getRawMany();
  }
} 

File: python-api/services/analytics_service.py
from sqlalchemy.orm import Session
from sqlalchemy import text
from typing import List

class AnalyticsService:
    def __init__(self, db: Session):
        self.db = db

    def get_orders_with_users(self, limit: int = 10, offset: int = 0):
        query = text("""
            SELECT o.id as order_id, o.user_id, u.name as user_name, u.email as user_email, 
                   u.city as user_city, o.total_amount, o.status, o.order_date,
                   COUNT(oi.id) as item_count
            FROM orders o
            JOIN users u ON o.user_id = u.id
            LEFT JOIN order_items oi ON o.id = oi.order_id
            GROUP BY o.id, o.user_id, u.name, u.email, u.city, o.total_amount, o.status, o.order_date
            ORDER BY o.id
            LIMIT :limit OFFSET :offset
        """)
        
        result = self.db.execute(query, {"limit": limit, "offset": offset})
        return result.fetchall()

    def get_user_order_summary(self, limit: int = 10, offset: int = 0):
        query = text("""
            SELECT u.id as user_id, u.name as user_name, u.email as user_email, 
                   COUNT(o.id) as total_orders,
                   COALESCE(SUM(o.total_amount), 0) as total_amount,
                   COALESCE(AVG(o.total_amount), 0) as average_order,
                   COALESCE(MAX(o.order_date), '1970-01-01'::timestamp) as last_order
            FROM users u
            LEFT JOIN orders o ON u.id = o.user_id
            GROUP BY u.id, u.name, u.email
            ORDER BY total_amount DESC
            LIMIT :limit OFFSET :offset
        """)
        
        result = self.db.execute(query, {"limit": limit, "offset": offset})
        return result.fetchall()

    def get_complex_analytics(self):
        query = text("""
            SELECT 
                p.category,
                COUNT(DISTINCT o.id) as total_orders,
                SUM(oi.quantity) as total_quantity,
                SUM(oi.price * oi.quantity) as total_revenue,
                AVG(oi.price) as avg_price,
                COUNT(DISTINCT u.id) as unique_customers,
                AVG(u.age) as avg_customer_age
            FROM products p
            JOIN order_items oi ON p.id = oi.product_id
            JOIN orders o ON oi.order_id = o.id
            JOIN users u ON o.user_id = u.id
            WHERE o.status = 'completed'
            GROUP BY p.category
            ORDER BY total_revenue DESC
        """)
        
        result = self.db.execute(query)
        return result.fetchall() 